### ISSUE TITLE
Reproducing BigGAN256 results

### ISSUE THREAD

**Issue by phymhan** [POSTECH-CVLab/PyTorch-StudioGAN#64]
Hi Mniguk,

Recently, I want to reproduce results of ProjGAN on ImageNet with config file **src/configs/ILSVRC2012/BigGAN256.json**. I use the command
```
python src/main.py -t -e -sync_bn -c src/configs/ILSVRC2012/BigGAN256.json --eval_type "valid"
```

From logging file **logs/IMAGENET/BigGAN256-train-2021_01_24_03_52_15.log** the FID at iteration 44000 is **46.32**, however, I get **43.87** when I run it. The difference is not negligible. I train BigGAN256 on 4 A100 gpus. I am wondering is this normal behavior? or is there anything I potentially did wrong?

The only differences of my experiment setting from using the command in README file are:
1. I use CenterCrop during training, but I think that would only make FID score worse.
2. I did not use -l option to save loading time.
3. -stat_otf flag is also off since it contradicts sync_bn.
4. I wrapped valid folder in another one since ILSVRC2012's valid folder is not organized in subfolders and causes problem in ImageFolder.

Thanks a lot in advance!

**Comment by MINGUKKANG** [comment]
Hi, 

Thank you so much for your interest in my repository.

The difference may be from two factors:

1. Just variance of training BigGAN. 

2. -stat_otf option: I think the difference may be from this option. Since -stat_otf can enhance the generation performances of GANs (CIFAR10 and Tiny ImageNet cases), I turned on the option to training BigGAN on ImageNet. I found, however, that using moving average statistics would give a better result on ImageNet generation experiment. So, I evaluated BigGAN on ImageNet using moving average statistics after the training (please refer to the training log).

Thank you!

**Comment by MINGUKKANG** [comment]
Please evaluate your model using -stat_otf option. It will clear your problem.

**Comment by phymhan** [comment]
Hi Minguk,

Thank you so much for your prompt reply! I will test with -stat_otf and update. A few quick questions:
1. If I understand correctly, you meant you turned on -stat_otf during training, how is this possible? since turning both on (stat_otf and sync_bn) doesn't pass config check.
2. You said "-stat_otf can enhance the generation performances", thus turning otf on presumably lowers the FID score. But my problem is I am getting lower FID and higher IS from my experiment than the provided log.
3. If I understand correctly, are you saying that in general "-stat_otf can enhance the generation performances", however, from your experience you find "using moving average statistics would give a better result on ImageNet"? Does this contradict Appendix C.1 in BigGAN paper or not? Sorry I don't have a lot experience in conducting these large scale experiments and not sure about what the common practice is when evaluating GAN models. Please correct me if I'm wrong.
4. From your experience, which evaluation protocol should I use: moving average, batch statistics, or standing statistics? and do you observe noticeable differences among them?

Thank you so much for your time!

Best,
Ligong

**Comment by phymhan** [comment]
Hi Minguk,

Just follow up my previous questions. I have tested my trained model with stat_otf on. The FID at iter 70000 is now 38.81 which matches the reported 39.36 in the log file much better. Previously from my training log, I got 36.20 without stat_otf. I think this 0.5 difference is within the variance between different runs. This also matches your saying "using moving average statistics would give a better result on ImageNet". Is this because when using small batch sizes (256), moving average statistics is more accurate while when batch size is large enough the benefit of using batch statistics becomes more evident?

I also observe that IS is less sensitive to bn options, the IS's at iter 70000 for my trained model without stat_otf, with stat_otf, and from the provided log are 19.17, 18.33, 18.40, respectively. Any explanation on this?

Also, I do want to know your suggestion on choosing bn options and evaluation protocols. Thanks a lot in advance!

Best,
Ligong

**Comment by MINGUKKANG** [comment]
Hi Ligong,

I think that our confusion is from the role of -stat_otf.

The -stat_otf option does not affect the training process of GAN. Instead, it lets the discriminator compute batch norms' statistics on the fly in the evaluation phase. 

[Examples]

1. Turning on -stat_otf leads the executed process not to synchronize statistics during training. Then, the process evaluates the model using non-synchronized statistics calculated on the fly.

2. Turning on -stat_otf and -sync_bn leads the executed process to synchronize statistics during training. And the process evaluates the model with synchronized statistics which are calculated on the fly.

3. Turning on -sync_bn only leads the executed process to synchronize statistics during training. However, the process evaluates the model using moving average statistics.

[My answer to your question]

1. Please refer to the examples above. When I train GANs on ImageNet, I thought that "-stat_otf" would give better results. I run all experiments with the "-stat_otf" option, but I noticed that using moving average statistics is more beneficial (It is too late to re-train all of my experiments...).

2. Please refer to the above.

3. It depends on the dataset you want to train. 

4.  I think "standing_statistics with proper steps" is the best choice. In the case of "-stat_otf" and "moving average statistics", I think it would be better to decide experimentally.

5. [Q] Is this because when using small batch sizes (256), moving average statistics is more accurate while when batch size is large enough the benefit of using batch statistics becomes more evident?
==> No, moving average statistics have shown a better result although batch_size is large enough (refer to BigGAN2048 experiments: 88.076(stat_otf) vs. 99.704(moving average)). I don't know the exact reason, but I suspect exponential moving average updated Generator (Gen_Copy) may cause this. Gen_Copy is a mischief-maker and let us go back and forth between heaven and hell.  

Thank you.

Best,

Minguk

**Comment by phymhan** [comment]
Hi Minguk,

Thank you so much for your reply! That's very clear. I think I'm confused about stat_otf because I didn't notice you have removed stat_otf and sync_bn assert in a recent commit after I forked your repo. Now it's clear. Thank you again!

best,
Ligong
